include/pluginlib/impl/filesystem_helper.hpp

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2017, Open Source Robotics Foundation, Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the Willow Garage, Inc. nor the names of its
+ *       contributors may be used to endorse or promote products derived from
+ *       this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/// Includes std::filesystem and aliases the namespace to `pluginlib::impl::fs`.
+/**
+ * If std::filesystem is not available the necessary functions are emulated.
+ */
+
+#ifndef SOMETHING_FILESYSTEM_HELPER
+#define SOMETHING_FILESYSTEM_HELPER
+
+#if defined(__has_include)
+#if __has_include(<filesystem>)
+# include <filesystem>
+
+namespace pluginlib
+{
+namespace impl
+{
+namespace fs = std::filesystem;
+}  // namespace impl
+}  // namespace pluginlib
+
+#define PLUGINLIB__IMPL__FILESYSYEM_HELPER__HAS_STD_FILESYSTEM
+#elif __has_include(<experimental/filesystem>)
+# include <experimental/filesystem>
+
+namespace pluginlib
+{
+namespace impl
+{
+namespace fs = std::experimental::filesystem;
+}  // namespace impl
+}  // namespace pluginlib
+
+#define PLUGINLIB__IMPL__FILESYSYEM_HELPER__HAS_STD_FILESYSTEM
+#endif
+#endif
+
+// The standard library does not provide it, so emulate it.
+#ifndef PLUGINLIB__IMPL__FILESYSYEM_HELPER__HAS_STD_FILESYSTEM
+
+#include <string>
+
+#ifdef _WIN32
+#define CLASS_LOADER_IMPL_OS_DIRSEP '\\'
+#else
+#define CLASS_LOADER_IMPL_OS_DIRSEP '/'
```

Comment:
Could filesystem give us the separator (using something like `fs::path::preferred_separator`) allowing us to not have to define them ourselves?

Reply:
This is the case where std::filesystem is not available...

Reply:
See: https://github.com/ros/pluginlib/pull/67/files#diff-a438fa6270a9a96b5deb3f4f2835e05cR96

Reply:
oh sorry, I read it too fast